File: data_structure/stack/src/StackDemo.js
// JavaScript
// データ構造: スタック (Stack)

class StackData {
    constructor() {
        this._data = [];
    }

    get() {
        return this._data;
    }

    get_index(item) {
        try {
            const index = this._data.indexOf(item);
            if (index === -1) {
                console.log(`ERROR: ${item} は範囲外です`);
                return -1;
            }
            return index;
        } catch (error) {
            console.log(`ERROR: ${item} は範囲外です`);
            return -1;
        }
    }

    get_value(index) {
        if (0 <= index && index < this._data.length) {
            return this._data[index];
        } else {
            console.log(`ERROR: ${index} は範囲外です`);
            return null;
        }
    }

    push(item) {
        this._data.push(item);
        return true;
    }

    pop() {
        if (!this.is_empty()) {
            this._data.pop();
            return true;
        } else {
            console.log(`ERROR: 空です`);
            return false;
        }
    }

    peek() {
        if (!this.is_empty()) {
            return this._data[this._data.length - 1];
        } else {
            return null;
        }
    }

    is_empty() {
        return this._data.length === 0;
    }

    size() {
        return this._data.length;
    }

    clear() {
        this._data = [];
        return true;
    }
}

function main() {
    console.log("Stack TEST -----> start");

    console.log("\nnew");
    const stack_data = new StackData();
    console.log(`  現在のデータ: ${JSON.stringify(stack_data.get())}`);

    console.log("\nis_empty");
    let output = stack_data.is_empty();
    console.log(`  出力値: ${output}`);

    console.log("\nsize");
    output = stack_data.size();
    console.log(`  出力値: ${output}`);

    console.log("\npush");
    const items_to_push = [10, 20, 30, 40];
    for (const item of items_to_push) {
        console.log(`  入力値: ${item}`);
        output = stack_data.push(item);
        console.log(`  出力値: ${output}`);
        console.log(`  現在のデータ: ${JSON.stringify(stack_data.get())}`);
    }

    console.log("\nsize");
    output = stack_data.size();
    console.log(`  出力値: ${output}`);

    console.log("\nis_empty");
    output = stack_data.is_empty();
    console.log(`  出力値: ${output}`);

    console.log("\npeek");
    output = stack_data.peek();
    console.log(`  出力値: ${output}`);

    console.log("\nget_index");
    let input = 30;
    console.log(`  入力値: ${input}`);
    output = stack_data.get_index(input);
    console.log(`  出力値: ${output}`);

    console.log("\nget_index");
    input = 50;
    console.log(`  入力値: ${input}`);
    output = stack_data.get_index(input);
    console.log(`  出力値: ${output}`);

    console.log("\npop");
    while (!stack_data.is_empty()) {
        output = stack_data.pop();
        console.log(`  出力値: ${output}`);
        console.log(`  現在のデータ: ${JSON.stringify(stack_data.get())}`);
    }

    console.log("\nis_empty");
    output = stack_data.is_empty();
    console.log(`  出力値: ${output}`);

    console.log("\nsize");
    output = stack_data.size();
    console.log(`  出力値: ${output}`);

    console.log("\npop");
    output = stack_data.pop();
    console.log(`  出力値: ${output}`);

    console.log("\npeek");
    output = stack_data.peek();
    console.log(`  出力値: ${output}`);

    console.log("\nStack TEST <----- end");
}

main();

File: data_structure/stack/src/StackDemo.kt
// Kotlin
// データ構造: スタック (Stack)

class StackData {
    private var _data = mutableListOf<Int>()

    fun get(): List<Int> {
        return _data
    }

    fun getIndex(item: Int): Int {
        val index = _data.indexOf(item)
        if (index == -1) {
            println("ERROR: $item は範囲外です")
        }
        return index
    }

    fun getValue(index: Int): Int? {
        return if (0 <= index && index < _data.size) {
            _data[index]
        } else {
            println("ERROR: $index は範囲外です")
            null
        }
    }
    
    fun push(item: Int): Boolean {
        _data.add(item)
        return true
    }
    
    fun pop(): Boolean {
        return if (!isEmpty()) {
            _data.removeAt(_data.size - 1)
            true
        } else {
            println("ERROR: 空です")
            false
        }
    }

    fun peek(): Int? {
        return if (!isEmpty()) {
            _data.last()
        } else {
            null
        }
    }
    
    fun isEmpty(): Boolean {
        return _data.isEmpty()
    }
    
    fun size(): Int {
        return _data.size
    }
    
    fun clear(): Boolean {
        _data.clear()
        return true
    }
}

fun main() {
    println("Stack TEST -----> start")

    println("\nnew")
    val stackData = StackData()
    println("  現在のデータ: ${stackData.get()}")

    println("\nis_empty")
    val isEmpty = stackData.isEmpty()
    println("  出力値: $isEmpty")

    println("\nsize")
    var output = stackData.size()
    println("  出力値: $output")

    println("\npush")
    val itemsToPush = listOf(10, 20, 30, 40)
    for (item in itemsToPush) {
        println("  入力値: $item")
        val pushResult = stackData.push(item)
        println("  出力値: $pushResult")
        println("  現在のデータ: ${stackData.get()}")
    }

    println("\nsize")
    output = stackData.size()
    println("  出力値: $output")

    println("\nis_empty")
    val isEmptyAfterPush = stackData.isEmpty()
    println("  出力値: $isEmptyAfterPush")

    println("\npeek")
    val peekResult = stackData.peek()
    println("  出力値: $peekResult")

    println("\nget_index")
    var input = 30
    println("  入力値: $input")
    val getIndexResult = stackData.getIndex(input)
    println("  出力値: $getIndexResult")

    println("\nget_index")
    input = 50
    println("  入力値: $input")
    val getIndexForMissingItem = stackData.getIndex(input)
    println("  出力値: $getIndexForMissingItem")

    println("\npop")
    while (!stackData.isEmpty()) {
        val popResult = stackData.pop()
        println("  出力値: $popResult")
        println("  現在のデータ: ${stackData.get()}")
    }

    println("\nis_empty")
    val isEmptyAfterPop = stackData.isEmpty()
    println("  出力値: $isEmptyAfterPop")

    println("\nsize")
    val sizeAfterPop = stackData.size()
    println("  出力値: $sizeAfterPop")

    println("\npop")
    val popEmptyResult = stackData.pop()
    println("  出力値: $popEmptyResult")

    println("\npeek")
    val peekEmptyResult = stackData.peek()
    println("  出力値: $peekEmptyResult")

    println("\nStack TEST <----- end")
}